Always check if default labels on node need to be updated in kubelet

mrIncompetent opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://github.com/kubernetes/community/blob/master/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:
Nodes join again but maybe OS/Arch/Instance-Type has changed in the meantime.
In this case the kubelet needs to check if the default labels are still correct and if not it needs to update them. 

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
Kubelet updates default labels if those are deprecated
```

smarterclayton committed:
Allow initialization of resources

Add support for creating resources that are not immediately visible to
naive clients, but must first be initialized by one or more privileged
cluster agents. These controllers can mark the object as initialized,
allowing others to see them.

Permission to override initialization defaults or modify an initializing
object is limited per resource to a virtual subresource "RESOURCE/initialize"
via RBAC.

Initialization is currently alpha.

jcbsmpsn committed:
Rotate kubelet client certificate.

Changes the kubelet so it bootstraps off the cert/key specified in the
config file and uses those to request new cert/key pairs from the
Certificate Signing Request API, as well as rotating client certificates
when they approach expiration.

gyliu513 committed:
Improved code coverage for pkg/kubelet/util.

The test coverage for pkg/kubelet/util.go increased from 45.1%
to 84.3%.

msau42 committed:
Add multizone PV tests

fabianofranz committed:
Add group alias names to discovery in registry

fabianofranz committed:
Add group alias names to discovery in API resources

fabianofranz committed:
Update generated

ailusazh committed:
Add SuccessfulMountVolume message to the events of pod

Commit:
Bump github.com/mitchellh/mapstructure

dixudx committed:
update gophercloud that fixed code format

mainly to include #265(https://github.com/gophercloud/gophercloud/pull/265)

ravigadde committed:
Added Bind method to Scheduler Extender

- only one extender can support the bind method
- if an extender supports bind, scheduler delegates the pod binding to the extender

mml committed:
Respect PDBs during GCE node upgrades.

Respect PDBs during node upgrades and add test coverage to the
ServiceTest upgrade test.  Modified that test so that we include pod
anti-affinity constraints and a PDB.

nbutton23 committed:
If ElbSecurityGroup is set then use it instead of creating another SG

DirectXMan12 committed:
[client-gen] Allow overriding the resource name

This commit allows client-gen to generate clients where the resource
name does not directly correspond to the type name.

The tag `// +resourceName=<name>` can be used to override the resource
name.

DirectXMan12 committed:
[client-gen] Support Read-Only APIs

This commit adds support for read-only APIs to client-gen.
If the tag `// +readonly=true` is used, only the {Get,List,Watch}
client methods will be generated.

Q-Lee committed:
Adding a metadata proxy addon to gce

DirectXMan12 committed:
Enable client-gen for k8s.io/metrics

This commit turns on client-gen for k8s.io/metrics.  Clients are
generated for `k8s.io/metrics/pkg/apis/metrics` (both internal and
v1alpha1).  `k8s.io/metrics/pkg/apis/custom_metrics` uses a bespoke
client due to the unique nature of its paths.

gyliu513 committed:
Toleration should be `notReady:NoExecute` in defaulttolerationseconds test.

Commit:
Clean up and restructure.

DirectXMan12 committed:
Fix HPA unit tests with new fake client

The new fake client properly represents the resource of `PodMetrics` as
"pods" and the resource of `NodeMetrics` as "nodes".  Previously, it
used "podmetricses" and "nodemetrics", respectively.

This fixes up `horizontal_test.go` and `replica_calc_test.go` to use the
new names.

Commit:
Added unit test for node operation in schedulercache.

DirectXMan12 committed:
Update generated clients for k8s.io/metrics

This commit updates the generated clients for k8s.io/metrics based on
the generators updated in the previous commits.

mml committed:
update-bazel.sh

Commit:
remove excpetion in verify-no-vendor-cycles.sh

Commit:
delete the useless gv

Signed-off-by: yupengzte <yu.peng36@zte.com.cn>

janetkuo committed:
Update DaemonSet API for rollback and history

1. Add revisionHistoryLimit (default 10), collisionCount, and validation code
2. Add daemonset-controller-hash label, and deprecate templateGeneration

Commit:
Include event messages in aggregated events

This changes the event aggregation behavior so that, when multiple events are
deduplicated, the aggregated event includes the message of the latest related
event.

This fixes an issue where the original event expires due to TTL, and the
aggregate event doesn't contain any useful message.

janetkuo committed:
Implement DaemonSet history logic in controller

1. Create controllerrevisions (history) and label pods with template
   hash for both RollingUpdate and OnDelete update strategy
2. Clean up old, non-live history based on revisionHistoryLimit
3. Remove duplicate controllerrevisions (the ones with the same template)
   and relabel their pods
4. Update RBAC to allow DaemonSet controller to manage
   controllerrevisions
5. In DaemonSet controller unit tests, create new pods with hash labels

Commit:
Some minor corrections in service controller

deads2k committed:
retry clientCA post start hook on transient failurs

mml committed:
Log more verbose when DisruptionController doesn't start.

To help with debugging #45706

janetkuo committed:
Implement kubectl rollout history and undo for DaemonSet

sttts committed:
apimachinery: move unversioned registration to metav1

janetkuo committed:
Test both strategies in all daemonSet controller unit tests

Commit:
Close file after os.Open()

dashpole committed:
trigger aggressive container garbage collection when under disk pressure

yguo0905 committed:
Support validating package versions in node conformance test

Q-Lee committed:
Add iptables lock-file mount to kube-proxy manifest

Q-Lee committed:
Add annotation for image policy webhook fail open.

mikedanese committed:
enable tls bootstrap in GCE/GKE

waseem committed:
Denote if a printer is generic.

This fixes #38779.

This allows us to avoid case in which printers.GetStandardPrinter
returns nil for both printer and err removing any potential panics that
may arise throughout kubectl commands.

Please see #38779 and #38112 for complete context.

Add comment explaining adding handlers to printers.HumanReadablePrinter
also remove an unnecessary conversion of printers.HumanReadablePrinter
to printers.ResourcePrinter.

sttts committed:
apimachinery: idempotent registration of unversioned types

smarterclayton committed:
Add priority and condition to column definitions

Allow the server to indicate additional information along with each row
and table.

whitlockjc committed:
add "install" package for admission API group

To properly register the types in the admission API group we need to
create an "install" package and wire it up.  This is required by the
webhook admission controller being developed as part of
https://github.com/kubernetes/community/pull/132

whitlockjc committed:
add generated sources for admission API group "install" package

janetkuo committed:
Autogen: run hack/update-all.sh

whitlockjc committed:
add generic webhook admission controller

As part of https://github.com/kubernetes/community/pull/132, thsi commit
adds a generic webhook admission controller.  This plugin allows for a
completely declarative approach for filtering/matching admission requests
and for matching admission requests, calls out to an external webhook for
handling admission requests.

whitlockjc committed:
add generated sources for webhook admission plugin

divyenpatel committed:
Adding vsphere Storage API Latency and Error Metrics support

fix bazel failure

Commit:
Add configuration options for encryption providers

Add location transformer, config for transformers

Location transformer helps choose the most specific transformer for
read/write operations depending on the path of resource being accessed.

Configuration allows use of --experimental-encryption-provider-config
to set up encryption providers. Only AEAD is supported at the moment.

Add new files to BUILD, AEAD => k8s-aes-gcm

Use group resources to select encryption provider

Update tests for configuration parsing

Remove location transformer

Allow specifying providers per resource group in configuration

Add IdentityTransformer configuration option

Fix minor issues with initial AEAD implementation

Unified parsing of all configurations

Parse configuration using a union struct

Run configuration parsing in APIserver, refactor parsing

More gdoc, fix minor bugs

Add test coverage for combined transformers

Use table driven tests for encryptionconfig

xilabao committed:
update default translation of annotations

Commit:
add test and fix typo in daemoncontroller

sjenning committed:
reset resultRun on pod restart

deads2k committed:
prevent illegal verb/name combinations in default policy rules

DirectXMan12 committed:
Add HPA status conditions to API types

This commit adds the new API status conditions to the API types.
The field exists as a field in autoscaling/v2alpha1, and is
round-tripped through an annotation in autoscaling/v1.

DirectXMan12 committed:
Make HPA controller set HPA status conditions

This commit causes the HPA controller to set a variety of status
conditions using the new `Status.Conditions` field of
autoscaling/v2alpha1.  These provide insight into the current state
of the HPA, and generally correspond to similar events being emitted.

janetkuo committed:
Verify histories and pods in DaemonSet e2e test

lavalamp committed:
Implement dynamic admission webhooks

Also fix a bug in rest client

xiangpengzhao committed:
Fix selfLinks of pods started from manifests

CaoShuFeng committed:
Fix doc about Verb for advanced audit feature

CaoShuFeng committed:
This change add nonResourceURL to kubectl auth cani

Commit:
Add RegisterFakeOnDelete to test federation object deletion

Commit:
Add federated service delete unit testcase

Commit:
Auto generated files

jingxu97 committed:
Add local storage (scratch space) allocatable support

This PR adds the support for allocatable local storage (scratch space).
This feature is only for root file system which is shared by kubernetes
componenets, users' containers and/or images. User could use
--kube-reserved flag to reserve the storage for kube system components.
If the allocatable storage for user's pods is used up, some pods will be
evicted to free the storage resource.

lavalamp committed:
generated changes

xilabao committed:
add policy_compact to rbac validation

smarterclayton committed:
Add Initializers to all admission control paths by default

xilabao committed:
add role/clusterrole to describe.go

xiangpengzhao committed:
Do not log the content of pod manifest if parsing fails.

Commit:
Fix some comments in dnsprovider

ajitak committed:
Configure NPD version through env variable

perotinus committed:
[Federation] Add a SchedulingAdapter that can extend the FederatedTypeAdapter and that provides hooks for scheduling objects into clusters.

tallclair committed:
Set up basic test configuration for AdavencedAuditing

DirectXMan12 committed:
Update kubectl to display HPA status conditions

This commit updates `kubectl describe` to display the new HPA
status conditions.  This should make it easier for users to discern
the current state of the HPA.

smarterclayton committed:
Grow signature for predicate attributes to include init status

DirectXMan12 committed:
Update generated autoscaling files

This commit updates the generated autoscaling files to be up-to-date
with the HPA status condition changes.

smarterclayton committed:
Refactor printers to support rendering as a Table

Return tables from the server.

Commit:
validation of subresources

tallclair committed:
Add AdvancedAuditing E2E test

smarterclayton committed:
Allow initialization when no authorizer present

Running without an authorizer is a valid configuration.

smarterclayton committed:
generated: client-go staging

cheftako committed:
Enable Dialer on the Aggregator

Centralize the creation of the dialer during startup.
Have the dialer then passed in to both APIServer and Aggregator.
Aggregator the sets the dialer on its Transport base.
This should allow the SSTunnel to be used but also allow the Aggregation
Auth to work with it.
Depending on Environment InsecureSkipTLSVerify *may* need to be set to
true.
Fixed as few tests to call CreateDialer as part of start-up.

sttts committed:
integration test: check API version onf Status object

smarterclayton committed:
Expose a default Table and partial output via Accept headers

All generic registries expose metadata output, and refactor endpoints to
allow negotiation to handle those responses. Add support for
PartialObjectMetadata being returned for objects as well.

Commit:
Fix inconsistenty in finding cni binaries
Fixes [#46476]

Signed-off-by: Abhinav Dahiya <abhinav.dahiya@coreos.com>

enxebre committed:
Improving test coverage for kubelet/kuberuntime.

gyliu513 committed:
Highlight nodeSelector when checking nodeSelector for Pod.

deads2k committed:
move labels to components which own the APIs

jingxu97 committed:
Add predicates check for local storage request

This PR adds the check for local storage request when admitting pods. If
the local storage request exceeds the available resource, pod will be
rejected.

Commit:
fix flake

jingxu97 committed:
Add EmptyDir Volume and local storage for container overlay Isolation

This PR adds two features:
1. add support for isolating the emptyDir volume use. If user
sets a size limit for emptyDir volume, kubelet's eviction manager
monitors its usage
and evict the pod if the usage exceeds the limit.
2. add support for isolating the local storage for container overlay. If
the container's overly usage exceeds the limit defined in container
spec, eviction manager will evict the pod.

monopole committed:
Add jregan to OWNERS for kubectl isolation work.

The kubectl decoupling project (#598) requires many BUILD edits.

Even relatively simple PR's involve many OWNER files, e.g. #46317 involves five.

We plan to script-generate some PRs, and those may involve _hundreds_ of BUILD files.

This project will take many PRs, and collecting all approvals for each
will be very time consuming.

Commit:
Rename gce_loadbalancer.go to gce_loadbalancer_external.go

Commit:
Rename gce_staticip.go to gce_addresses.go

mbohlool committed:
Enable OpenAPI definition generation for apiregistration

yguo0905 committed:
Update semver to 3.5.0

Commit:
nonnamespaced tag

Commit:
generated

smarterclayton committed:
Load initializers from dynamic config

Handle failure cases on startup gracefully to avoid causing cascading
errors and poor initialization in other components. Initial errors from
config load cause the initializer to pause and hold requests. Return
typed errors to better communicate failures to clients.

Add code to handle two specific cases - admin wants to bypass
initialization defaulting, and mirror pods (which want to bypass
initialization because the kubelet owns their lifecycle).

xilabao committed:
1. Fix create volume of CertificatesDir.
2. Replace "manifests" to kubeadmconstants.ManifestsDirName

gyliu513 committed:
Move tolerations to PodSpec for calico-node.yaml.

gyliu513 committed:
Move tolerations to PodSpec for ip-masq-agent.yaml.

Commit:
Fix for-loop and err definition

tnozicka committed:
Fix standardFinalizers - add missing metav1.FinalizerDeleteDependents

gyliu513 committed:
Added a new test case for daemoncontroller.

This patch added a new test case of daemonSet with node selector,
matching some nodes, and launch pods on all the nodes.

CaoShuFeng committed:
add some unit tests for "kubectl create clusterrole"

Commit:
Add event exporter deployment to the fluentd-gcp addon

CaoShuFeng committed:
Fix unit test for kubectl create role

x1957 committed:
adds log when --kubeconfig with wrong config

thockin committed:
Proxy: comments on --masquerade-all flag

thockin committed:
Proxy: comments around ClusterCIDR use

thockin committed:
'Global' -> 'Cluster' for traffic policy

tallclair committed:
Generate protobuf for the audit API

mikedanese committed:
promote tls-bootstrap to beta

a-robinson committed:
Support custom domains in the cockroachdb example's init container

This switches from using v0.1 of the peer-finder image to a version that
includes https://github.com/kubernetes/contrib/pull/2013

While I'm here, switch the version of cockroachdb from 1.0 to 1.0.1

mikedanese committed:
change kubemark image project to match new cos image project

Commit:
Log out from multiple portals with iscsi storage plugin

When using iscsi storage with multiple target portal (TP)
addresses and multipathing the volume manager logs on to
the IQN for all portal addresses, but when a pod gets
destroyed the volume manager only logs out for the primary
TP and sessions for another TPs are always remained.

This patch adds methods to store and load iscsi disk
configrations, then uses the stored config at DetachDisk
path.

Fix #45394

deads2k committed:
stop special casing the loopback connection for aggregator

tallclair committed:
Capture better debug logs on AppArmor test failure

deads2k committed:
add protobuf for CRD

janetkuo committed:
Add controllerrevisions as valid resource in kubectl help

tallclair committed:
Instrument advanced auditing

shyamjvs committed:
Enable DefaultTolerationSeconds and PodPreset admission plugins for kubemark

mbohlool committed:
Update OpenAPI spec

mbohlool committed:
Update Bazel

Commit:
do not allow subresources in initializer rules

Commit:
generated

cblecker committed:
Exclude dockerized verify patterns

tallclair committed:
Fix AppArmor test for docker 1.13

Commit:
update gengo godep

Commit:
Add some initial resource limits to the ip-masq-agent.

pipejakob committed:
kubeadm: don't duplicate master taint if it already exists.

This helps enable a graceful upgrade/downgrade process between 1.6.x and
1.7.x kubeadm clusters (although no guarantees outside of that range) by
doing:

  $ kubeadm init --kubernetes-version <version> --skip-preflight-checks

Without this change, the command fails with an error that the node taint
is duplicated.

This is part of https://github.com/kubernetes/kubeadm/issues/278

Commit:
generated defaults files

generated bazel

mbohlool committed:
Remove unused servePath from GetOperationIDAndTags and GetDefinitionName

mbohlool committed:
Separate Build and Serving parts of OpenAPI spec handler

mbohlool committed:
Update proto

Commit:
Delete meaningless check

xingzhou committed:
While deleting a namespace, the TPR instances under this ns should be deleted.

While deleting a namespace, the TPR instances under this ns should be deleted.

Commit:
Improve code coverage for kubelet

wasylkowski-a committed:
Added missing documentation to NodeInstanceGroup.

CaoShuFeng committed:
empty audit policy file is legal configuration

Empty audit policy file or policy file contains only comments means
using default audit level for all requests.

mbohlool committed:
Aggregate OpenAPI specs

n-marton committed:
Added node to persistent-volume-binder clusterrole

sttts committed:
apiserver: avoid resolving 'localhost'

The golang stdlib resolve localhost first via the external DNS server,
not via /etc/hosts. If your DNS resolve localhost.<search-domain>, the
API server won't start.

liggitt committed:
Pre-generate SNI test certs

deads2k committed:
complete the controller context for init funcs

mikedanese committed:
allow sarapprover to create subject access reviews

grodrigues3 committed:
Added sig leads alias to OWNERS_ALIAS

Commit:
Update the fluentd-gcp image

shyamjvs committed:
Add KernelDeadlock condition to hollow NPD

dashpole committed:
add semicolon

verult committed:
PD e2e test: Ready node check now uses the most up-to-date node count.

MrHohn committed:
Bump cluster-proportional-autoscaler to 1.1.2

Commit:
Refactor compute API calls

Commit:
Modify external LB logic

Commit:
Add internal LB logic

Commit:
Hook external & internal lb together

Commit:
Update test utilities & build file

vishh committed:
update cos kernel sha for node e2e GPU installer

Signed-off-by: Vishnu kannan <vishnuk@google.com>

dashpole committed:
update cadvisor godeps

ixdy committed:
bazel: base kube-proxy image on debian-iptables instead of busybox

ash2k committed:
Fix round-trip of Unstructured.OwnerReferences

Previously setOwnerReference was storing pointers but
extractOwnerReference is expecting pointer fields as plain values.
Fixes #46817

pipejakob committed:
kubeadm: Only create bootstrap configmap if exists.

The fact that this method was not idempotent was breaking kubeadm
upgrades.

CaoShuFeng committed:
update copyed doc for advanced audit

doc for WithAudit is copyed from WithLegacyAudit, it's out of date.
This change update doc for these two functions.

sttts committed:
apiserver: return BadRequest 400 for invalid query params

Commit:
Always check if default labels on node need to be updated in kubelet

dashpole committed:
update prometheus dependency for staging

mtaufen committed:
Fix cAdvisorPort, 0 is a valid option

feiskyer commented:
@k8s-bot ok to test

supereagle commented on pkg/kubelet/kubelet_node_status.go:
```diff
@@ -156,6 +157,32 @@ func (kl *Kubelet) tryRegisterWithApiServer(node *v1.Node) bool {
 }
 
+//updateDefaultLabels will set the default labels on the node
+func (kl *Kubelet) updateDefaultLabels(initialNode, existingNode *v1.Node) (needsUpdate bool) {
```
Please keep the code style consistent, no need to define the return variable here.

supereagle commented on pkg/kubelet/kubelet_node_status.go:
```diff
@@ -156,6 +157,32 @@ func (kl *Kubelet) tryRegisterWithApiServer(node *v1.Node) bool {
+	for _, label := range defaultLabels {
+		if existingNode.Labels[label] != initialNode.Labels[label] {
+			existingNode.Labels[label] = initialNode.Labels[label]
+			needsUpdate = true
```
Directly return true.

mrIncompetent replied:
If it would directly return true, only the first update would be updated

It needs to iterate over all default labels and update them if needed

supereagle commented on pkg/kubelet/kubelet_node_status.go:
```diff
@@ -156,6 +157,32 @@ func (kl *Kubelet) tryRegisterWithApiServer(node *v1.Node) bool {
+		}
+	}
+
+	return needsUpdate
```
Directly return false.

Commit:
Add fatal message

Commit:
Use new kubelet apis pkg for labels

shyamjvs committed:
Avoid double printing output of gcloud commands in kubemark

madhusudancs committed:
Do not delete PVs with --all, instead delete them selectively.

PV is a non-namespaced resource. Running `kubectl delete pv --all`, even
with `--namespace` is going to delete all the PVs in the cluster. This
is a dangerous operation and should not be deleted this way.

Instead we now retrieve the PVs bound to the PVCs in the namespace we
are deleteing and delete only those PVs.

Fixes issue #46380.

madhusudancs committed:
Delete cluster role and their bindings federated clusters.

This is part of the namespace deletion big hammer. `kubefed join` not
just creates federation-system namespace, but also cluster role and
cluster role bindings in the joining clusters. Sometimes unjoin fails
to delete them. So we use a big hammer here to delete them.

This smells like a real problem in kubefed and needs investigation.
This is a short term fix to unblock the submit queue.

madhusudancs committed:
Delete federation system namespace from all the federated clusters.

This is a big hammer. `kubefed join` creates federation-system namespace
in the joining clusters if they don't already exist. This namespace
usually exists in the host cluster and hence cannot be deleted while
unjoining. So in order to be safe, we don't delete the federation-system
namespace from any federated cluster while unjoining them. This causes
a problem in our test environment if certain resources are left in the
namespace. Therefore we are deleting all federation-system namespace in
all the clusters.

Commit:
bkpPortal should be initialized beforehand

This patch is a follow up patch for the PR#46239.
The bkpPortal in DetachDisk() path should be initialized
before using it.

deads2k committed:
deflake CRD watch tests

liggitt committed:
Check live pod on cache miss before forbidding node deletion

liggitt committed:
Avoid * in filenames

smarterclayton committed:
Remove Initializers from federation

deads2k committed:
make the health check wait for ready apiservices

mikedanese committed:
save kubelet bootstrap certs on GCI

Commit:
Fix error when the existing node has an uninitialized labels map

Commit:
Update tests

Commit:
Move named return parameter to local variable

mrIncompetent commented:
/retest

Commit:
Update label package
